app/models/workflow_step.rb

```diff
@@ -39,11 +40,21 @@ def set_completed_at
     self.completed_at = Time.now
   end
 
+  # Check if step's status is a completed one *and* that there is no completed_at value
+  def completed_and_not_already_completed?
+    completed? && !already_completed?
+  end
+
   ##
   # indicate if this step is marked as completed
   # @return [boolean]
   def completed?
-    COMPLETED_STATES.include? status
+    COMPLETED_STATES.include?(status)
+  end
+
+  # Check if there is already a completed_at date
+  def already_completed?
```

Comment:
and then this could be `maybe_set_completed` which would have all the conditions. (e.g. `completed? && !completed_at`)

Reply:
ok, I like it. will do.